pkg/server/sotw/v3/server.go

```diff
@@ -183,35 +164,53 @@ func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryReq
 				}
 			}
 
-			if lastResponse, ok := lastDiscoveryResponses[req.TypeUrl]; ok {
-				if lastResponse.nonce == "" || lastResponse.nonce == nonce {
-					// Let's record Resource names that a client has received.
-					streamState.SetKnownResourceNames(req.TypeUrl, lastResponse.resources)
-				}
+			typeURL := req.GetTypeUrl()
+			// State cannot be modified until any potential watch is closed
+			state, ok := streamStates[typeURL]
+			if !ok {
+				// We don't have a current state for this type, create one
+				state = stream.NewStreamState(len(req.ResourceNames) == 0, nil)
+			} else if nonce != state.LastResponseNonce() {
+				// The request does not match the last response we sent.
+				// The protocol is a bit unclear in this case, but currently we discard such request and wait
+				// for the client to acknowledge the previous response.
+				// This is unclear how this handles cases where a response would be missed as any subsequent request will be discarded.
+
+				// We can continue here as the watch list hasn't changed so we don't need to recompute the watches select
+				continue
 			}
 
-			typeURL := req.GetTypeUrl()
 			responder := make(chan cache.Response, 1)
 			if w, ok := watches.responders[typeURL]; ok {
-				// We've found a pre-existing watch, lets check and update if needed.
-				// If these requirements aren't satisfied, leave an open watch.
-				if w.nonce == "" || w.nonce == nonce {
-					w.close()
-
-					watches.addWatch(typeURL, &watch{
-						cancel:   s.cache.CreateWatch(req, streamState, responder),
-						response: responder,
-					})
+				// If we had an open watch, close it to make sure we don't end up sending a cache response while we update the state of the request
+				w.close()
```

Comment:
The race mentioned here is currently present in the code if we receive a new request while we are building a response in the cache (race on KnownResourceNames)